source/common/stats/utility.cc

```diff
@@ -125,6 +125,62 @@ TextReadout& textReadoutFromStatNames(Scope& scope, const StatNameVec& elements,
   return scope.textReadoutFromStatNameWithTags(StatName(joined.get()), tags);
 }
 
+std::vector<ParentHistogram::Bucket>
+interpolateHistogramBuckets(uint32_t max_buckets, uint32_t num_src_buckets,
+                            std::function<ParentHistogram::Bucket()> next_bucket) {
+  if (max_buckets == InterpolateRetainAllBuckets) {
+    max_buckets = num_src_buckets;
+  }
+  const uint32_t num_buckets = std::min(max_buckets, num_src_buckets);
+  uint32_t num_src_buckets_per_bucket = 1;
+
+  // Determine how we will map source buckets to destination buckets.
+  // Consider 3 cases:
+  //  1. if we have 15 source buckets and 5 destination buckets we'll
```

Comment:
Depending on the distribution of the histogram, this may not be a good algorithm for reducing bucket count. It is often preferable to keep more buckets near the upper/lower limits, and fewer buckets in the middle. 

Can we get away with not limiting the buckets at all, and just using whatever the histogram library gives us? I believe it is a finite number of buckets, and not excessively large.

Reply:
I'm not sure what the maximum is from the histogram library, but I couldn't get it to give that many buckets, and felt that based on my experiments if I set 64 buckets that would generally not require any interpolation. The interpolation is just defensive to protect against a future version of circlhist, or a certain data pattern that I have not tried, resulting in a lot more buckets.

I agree with you putting the remainder at the beginning is not necessarily best but I'm not sure why another distribution would be better, and so I opted for what would be easiest :)

Reply:
I'd say for now, just add an `ENVOY_BUG(buckets <= 100)` or something and remove the interpolation code entirely.

Reply:
That's probably OK. We could also decide on that at call-sites by calling getDetail(MAX_BUCKETS + 1) and then asserting at call-site we got <= MAX_BUCKETS.

Reply:
@ggreenway : Well I tried this and I don't think I like it :) I was thinking there wouldn't be that many buckets but there can be, and they need interpolation. E.g. this test gives me 180 buckets:
```
  ScopeSharedPtr scope1 = store_->createScope("scope1.");
  Histogram& histogram = scope_.histogramFromString("histogram", Histogram::Unit::Unspecified);
  store_->mergeHistograms([]() -> void {});
  ASSERT_EQ(1, store_->histograms().size());
  ParentHistogramSharedPtr parent_histogram = store_->histograms()[0];

  for (uint32_t i = 0; i < 1000; ++i) {
    uint64_t value = i * 1000;
    EXPECT_CALL(sink_, onHistogramComplete(Ref(histogram), value));
    histogram.recordValue(value);
  }
  store_->mergeHistograms([]() -> void {});
  uint32_t total_buckets = parent_histogram->detailedTotalBuckets().size();
  EXPECT_LT(100, total_buckets);
```
The actual number of buckets I see is 190.

I think your main concern with the interpolation was that it wouldn't spread the way you might expect when NumBuckets is not an exact multiple of MaxBuckets. But I don't think it's likely to be that big a problem in practice. The interpolation algorithm makes the bucket wider and combines the counts, so data isn't really lost; just resolution. And if the number of buckets is much higher due to extreme dynamic range then the unevenly spread remainder will be much smaller in magnitude than the evenly spread excess.  E.g. if we have 190 buckets from circlhist and want to display 32 buckets, we'll get 6:1 bucket combinations for the 30 buckets, and for 5:1 bucket combinations for 2.

We could iterate a bit on cleaning up the distribution if you like, but from a UI perspective I'm going to want to interpolate at some point or the graphics will bleed due to the text annotations.

One option is to have /stats?format=json&histogram_bucket_mode=detailed&num_buckets=unlimited could give full resolution (possibly allocating large temp string buffers), but we could use &num_buckets=32 which would be applied by the ajax rendering layer prototyped in https://github.com/envoyproxy/envoy/pull/26472 .

Reply:
OK I checked in a revised model where I keep a min_value and width for each bucket (exactly what I get from circlhist) which is actually a bit cleaner from a merge-semantics perspective. We make the buckets wider and combine the counts.

Happy to iterate on the exact distribution algorithm but would you consider that as being something we could do in a follow-up?

Reply:
If we just want to render geometry it's probably OK, but I wanted to annotate with actual values as text. I think 30-40 is about that max we'd want to interpolate to.

My thinking is that circlhist is reasonably good at picking buckets logarithmically needed for a dynamic range, and that incrementally changing them by combining neighbors until you get the number of buckets you want should avoid being misleading. I think changing from just a bucket 'value' to a range (min,width) makes me feel like we are not distorting the data with this transform.

But in any case this isn't changing the existing behavior; this would (at least for now) only be used for graphical rendering. We could then iterate on how we combine them.

Another option is to have the C++ always yield all the buckets and sort out interpolation in JS.

Reply:
I think we don't get much value from the dynamic impl right now in terms of detailed buckets, but I suspect it currently helps us provide accurate p95, p99, p99.5 etc across a wide range of data values. And with this series of PRs we'd get detailed bucketing for a single proxy via its admin endpoint. I think this coercion I proposed would rarely be activated in practice; it'd only come into play with a very wide dynamic range of actual detail.

We could implement the coercion in C++ but let the endpoint query-params determine whether to coerce at all or what the max bucket-count should be.  I think in practice we'll get less than 30 buckets; but I don't have actual data for this yet.

However I'm warming up to moving the coercion to JS as that way a UI could easily be added to show the detail within a merged bucket. So I'll remove the coercion from this PR. I'll move status to 'draft' for now and and re-open soon. Thanks for working through this.